File: leetcode/problems/python3/3614. Process String with Special Operations II/solve.py
class Solution:
    def processStr(self, s: str, k: int) -> str:
        L = 0
        for c in s:
            if c == "*":
                L = max(L - 1, 0)
            elif c == "#":
                L *= 2
            elif c == "%":
                pass
            else:
                L += 1

        if k > L - 1:
            return "."

        for c in reversed(s):
            if c == "*":
                L += 1
            elif c == "#":
                L //= 2
                if k >= L:
                    k -= L
            elif c == "%":
                k = L - 1 - k
            else:
                if k == L - 1:
                    return c
                else:
                    L -= 1

        return "."

File: leetcode/problems/go/3614. Process String with Special Operations II/solve.go
package main

func processStr(s string, k int64) byte {
	var L int64
	for _, c := range s {
		switch c {
		case '*':
			L = max(L-1, 0)
		case '#':
			L *= 2
		case '%':
			// do nothing
		default:
			L++
		}
	}

	if k > L-1 {
		return '.'
	}
	for i := len(s) - 1; i >= 0; i-- {
		c := s[i]
		switch c {
		case '*':
			L++
		case '#':
			L /= 2
			if k > L-1 {
				k -= L
			}
		case '%':
			k = L - 1 - k
		default:
			if k == L-1 {
				return byte(c)
			} else {
				L--
			}
		}
	}

	return '.'
}
